Question 15.2

In the videos, we saw the “diet problem”. (The diet problem is one of the first large-scale optimization
problems to be studied in practice. Back in the 1930’s and 40’s, the Army wanted to meet the nutritional
requirements of its soldiers while minimizing the cost.) In this homework you get to solve a diet problem with real data. The data is given in the file diet.xls. 

1.	Formulate an optimization model (a linear program) to find the cheapest diet that satisfies the maximum and minimum daily nutrition constraints, and solve it using PuLP.  Turn in your code and the solution. (The optimal solution should be a diet of air-popped popcorn, poached eggs, oranges, raw iceberg lettuce, raw celery, and frozen broccoli. UGH!)
2.	Please add to your model the following constraints (which might require adding more variables) and solve the new model:
a.	If a food is selected, then a minimum of 1/10 serving must be chosen. (Hint: now you will need two variables for each food i: whether it is chosen, and how much is part of the diet. You’ll also need to write a constraint to link them.)
b.	Many people dislike celery and frozen broccoli. So at most one, but not both, can be selected.
c.	To get day-to-day variety in protein, at least 3 kinds of meat/poultry/fish/eggs must be selected. [If something is ambiguous (e.g., should bean-and-bacon soup be considered meat?), just call it whatever you think is appropriate – I want you to learn how to write this type of constraint, but I don’t really care whether we agree on how to classify foods!]

If you want to see what a more full-sized problem would look like, try solving your models for the file diet_large.xls, which is a low-cholesterol diet model (rather than minimizing cost, the goal is to minimize cholesterol intake).  I don’t know anyone who’d want to eat this diet – the optimal solution includes dried chrysanthemum garland, raw beluga whale flipper, freeze-dried parsley, etc. – which shows why it’s necessary to add additional constraints beyond the basic ones we saw in the video!
	[Note: there are many optimal solutions, all with zero cholesterol, so you might get a different one.  It probably won’t be much more appetizing than mine.]



SOLUTION:
    
Firstly, lets install the Pulp package from python. Then, we will build a model to solve the first question to apply the basic minimum/maximum nutrient constraints to find an optimal dietary solution.

Later, we will add specfic constraints as per the requirements in Question 2 to find the optimal solution

In [26]:
#pip install pulp
from pulp import *
import pandas as pd

1) Formulate an optimization model (a linear program) to find the cheapest diet that satisfies the maximum and minimum daily nutrition constraints, and solve it using PuLP. Turn in your code and the solution. (The optimal solution should be a diet of air-popped popcorn, poached eggs, oranges, raw iceberg lettuce, raw celery, and frozen broccoli. UGH!)

In [27]:
# Reading the data from the input excel file
data = pd.read_excel("diet.xls")

# The last 3 rows contain blank, minimum and maximum limits to each of the food nutritens which can be exlcuded 
food_data = data[0:64]

# Creating a list of all the food available
food_list = food_data.values.tolist()

# Creating a list for food names for further use in the optimization model
food_names = [x[0] for x in food_list]

# Creating a dictionary to hold food names along with the other nutrients in the data set for further processing
food_costs = dict([(x[0], float(x[1])) for x in food_list])
food_cals = dict([(x[0], float(x[3])) for x in food_list])
food_serving = dict([(x[0], x[2]) for x in food_list])
food_choles = dict([(x[0], float(x[4])) for x in food_list])
food_totfat = dict([(x[0], float(x[5])) for x in food_list])
food_sodium = dict([(x[0], float(x[6])) for x in food_list])
food_carbs = dict([(x[0], float(x[7])) for x in food_list])
food_fiber = dict([(x[0], float(x[8])) for x in food_list])
food_protein = dict([(x[0], float(x[9])) for x in food_list])
food_vitA = dict([(x[0], float(x[10])) for x in food_list])
food_vitC = dict([(x[0], float(x[11])) for x in food_list])
food_calcium = dict([(x[0], float(x[12])) for x in food_list])
food_iron = dict([(x[0], float(x[13])) for x in food_list])

################################  BUILDING THE OPTIMIZATION MODEL ###############################
## Defining the optimization problem with a name and mentioning if it is maximizing or minimizing
opt_problem = LpProblem("Diet_Problem",LpMinimize)

## Defining the primary variables for the optimization model with a natural lower bound of 0
amount_var = LpVariable.dicts("Amounts",food_names, 0)

## Adding the objective function
opt_problem += lpSum([food_costs[i] * amount_var[i] for i in food_names]), 'total_cost'

## Add THE MINIMUM/ Maximum daily nutrient constains to the model to check the optimal diet solution
#Calories
opt_problem += lpSum([food_cals[i] * amount_var[i] for i in food_names]) >= 1500, 'min_cals'
opt_problem += lpSum([food_cals[i] * amount_var[i] for i in food_names]) <= 2500, 'max_cals'
#Cholestrol
opt_problem += lpSum([food_choles[i] * amount_var[i] for i in food_names]) >= 30, 'min_chol'
opt_problem += lpSum([food_choles[i] * amount_var[i] for i in food_names]) <= 240, 'max_chol'
#Total Fat
opt_problem += lpSum([food_totfat[i] * amount_var[i] for i in food_names]) >= 20, 'min_fat'
opt_problem += lpSum([food_totfat[i] * amount_var[i] for i in food_names]) <= 70, 'max_fat'
#Sodium
opt_problem += lpSum([food_sodium[i] * amount_var[i] for i in food_names]) >= 800, 'min_sodium'
opt_problem += lpSum([food_sodium[i] * amount_var[i] for i in food_names]) <= 2000, 'max_sodium'
#Carbohydrates
opt_problem += lpSum([food_carbs[i] * amount_var[i] for i in food_names]) >= 130, 'min_carbs'
opt_problem += lpSum([food_carbs[i] * amount_var[i] for i in food_names]) <= 450, 'max_carbs'
#Fiber
opt_problem += lpSum([food_fiber[i] * amount_var[i] for i in food_names]) >= 125, 'min_fiber'
opt_problem += lpSum([food_fiber[i] * amount_var[i] for i in food_names]) <= 250, 'max_fiber'
#Protein
opt_problem += lpSum([food_protein[i] * amount_var[i] for i in food_names]) >= 60, 'min_protein'
opt_problem += lpSum([food_protein[i] * amount_var[i] for i in food_names]) <= 100, 'max_protein'
#VitaminA
opt_problem += lpSum([food_vitA[i] * amount_var[i] for i in food_names]) >= 1000, 'min_vitA'
opt_problem += lpSum([food_vitA[i] * amount_var[i] for i in food_names]) <= 10000, 'max_vitA'
#VitaminC
opt_problem += lpSum([food_vitC[i] * amount_var[i] for i in food_names]) >= 400, 'min_vitC'
opt_problem += lpSum([food_vitC[i] * amount_var[i] for i in food_names]) <= 5000, 'max_vitC'
#Calcium
opt_problem += lpSum([food_calcium[i] * amount_var[i] for i in food_names]) >= 700, 'min_calcium'
opt_problem += lpSum([food_calcium[i] * amount_var[i] for i in food_names]) <= 1500, 'max_calcium'
#Iron
opt_problem += lpSum([food_iron[i] * amount_var[i] for i in food_names]) >= 10, 'min_iron'
opt_problem += lpSum([food_iron[i] * amount_var[i] for i in food_names]) <= 40, 'max_iron'

In [29]:
#### Solving the Optimizer
opt_problem.solve()

1

In [31]:
#Viewing the optimal solution with this model
varsDictionary = {}

for v in opt_problem.variables():
    varsDictionary[v.name] = v.varValue
    
print(varsDictionary)

{'Amounts_2%_Lowfat_Milk': 0.0, 'Amounts_3.3%_Fat,Whole_Milk': 0.0, 'Amounts_Apple,Raw,W_Skin': 0.0, 'Amounts_Apple_Pie': 0.0, 'Amounts_Bagels': 0.0, 'Amounts_Banana': 0.0, 'Amounts_Beanbacn_Soup,W_Watr': 0.0, 'Amounts_Bologna,Turkey': 0.0, 'Amounts_Butter,Regular': 0.0, "Amounts_Cap'N_Crunch": 0.0, 'Amounts_Carrots,Raw': 0.0, 'Amounts_Celery,_Raw': 52.64371, 'Amounts_Cheddar_Cheese': 0.0, 'Amounts_Cheerios': 0.0, 'Amounts_Chicknoodl_Soup': 0.0, 'Amounts_Chocolate_Chip_Cookies': 0.0, "Amounts_Corn_Flks,_Kellogg'S": 0.0, 'Amounts_Couscous': 0.0, 'Amounts_Crm_Mshrm_Soup,W_Mlk': 0.0, 'Amounts_Frankfurter,_Beef': 0.0, 'Amounts_Frozen_Broccoli': 0.25960653, 'Amounts_Frozen_Corn': 0.0, 'Amounts_Grapes': 0.0, 'Amounts_Ham,Sliced,Extralean': 0.0, 'Amounts_Hamburger_W_Toppings': 0.0, 'Amounts_Hotdog,_Plain': 0.0, 'Amounts_Kielbasa,Prk': 0.0, 'Amounts_Kiwifruit,Raw,Fresh': 0.0, 'Amounts_Lettuce,Iceberg,Raw': 63.988506, 'Amounts_Macaroni,Ckd': 0.0, 'Amounts_Malt_O_Meal,Choc': 0.0, 'Amounts_New_E_

We observe that the optimal solution is the same as provided in the question. The optimal diet pattern would be

Celery,_Raw  (52.64371 servings)
Frozen_Broccoli  (0.25960653 servings)
Lettuce,Iceberg,Raw  (63.988506 servings)
Oranges (2.2929389 servings)
Poached_Eggs (0.14184397 servings)
Popcorn,Air_Popped (13.869322 servings)

2) Please add to your model the following constraints (which might require adding more variables) and solve the new model: a. If a food is selected, then a minimum of 1/10 serving must be chosen. (Hint: now you will need two variables for each food i: whether it is chosen, and how much is part of the diet. You’ll also need to write a constraint to link them.) b. Many people dislike celery and frozen broccoli. So at most one, but not both, can be selected. c. To get day-to-day variety in protein, at least 3 kinds of meat/poultry/fish/eggs must be selected. [If something is ambiguous (e.g., should bean-and-bacon soup be considered meat?), just call it whatever you think is appropriate – I want you to learn how to write this type of constraint, but I don’t really care whether we agree on how to classify foods!]

In [46]:
# We already have as a variable for food names for further use in the optimization model
food_names = [x[0] for x in food_list]

# Creating a dictionary to hold food names along with the other nutrients in the data set for further processing
food_costs = dict([(x[0], float(x[1])) for x in food_list])
food_cals = dict([(x[0], float(x[3])) for x in food_list])
food_serving = dict([(x[0], x[2]) for x in food_list])
food_choles = dict([(x[0], float(x[4])) for x in food_list])
food_totfat = dict([(x[0], float(x[5])) for x in food_list])
food_sodium = dict([(x[0], float(x[6])) for x in food_list])
food_carbs = dict([(x[0], float(x[7])) for x in food_list])
food_fiber = dict([(x[0], float(x[8])) for x in food_list])
food_protein = dict([(x[0], float(x[9])) for x in food_list])
food_vitA = dict([(x[0], float(x[10])) for x in food_list])
food_vitC = dict([(x[0], float(x[11])) for x in food_list])
food_calcium = dict([(x[0], float(x[12])) for x in food_list])
food_iron = dict([(x[0], float(x[13])) for x in food_list])

################################  BUILDING THE OPTIMIZATION MODEL ###############################
## Defining the optimization problem with a name and mentioning if it is maximizing or minimizing
opt_problem2 = LpProblem("Diet_Problem",LpMinimize)

## Defining the primary variables for the optimization model with a natural lower bound of 0
amount_var = LpVariable.dicts("Amounts",food_names, 0)

# Let's create a binary variable that will indicate if a particular food was used in the diet or not.
binaryVars = LpVariable.dicts("ChosenYOrN", food_names,0,1, LpBinary)

## Adding the objective function
opt_problem2 += lpSum([food_costs[i] * amount_var[i] for i in food_names]), 'total_cost'

## Add THE MINIMUM/ Maximum daily nutrient constains to the model to check the optimal diet solution (basic constraints)
#Calories
opt_problem2 += lpSum([food_cals[i] * amount_var[i] for i in food_names]) >= 1500, 'min_cals'
opt_problem2 += lpSum([food_cals[i] * amount_var[i] for i in food_names]) <= 2500, 'max_cals'
#Cholestrol
opt_problem2 += lpSum([food_choles[i] * amount_var[i] for i in food_names]) >= 30, 'min_chol'
opt_problem2 += lpSum([food_choles[i] * amount_var[i] for i in food_names]) <= 240, 'max_chol'
#Total Fat
opt_problem2 += lpSum([food_totfat[i] * amount_var[i] for i in food_names]) >= 20, 'min_fat'
opt_problem2 += lpSum([food_totfat[i] * amount_var[i] for i in food_names]) <= 70, 'max_fat'
#Sodium
opt_problem2 += lpSum([food_sodium[i] * amount_var[i] for i in food_names]) >= 800, 'min_sodium'
opt_problem2 += lpSum([food_sodium[i] * amount_var[i] for i in food_names]) <= 2000, 'max_sodium'
#Carbohydrates
opt_problem2 += lpSum([food_carbs[i] * amount_var[i] for i in food_names]) >= 130, 'min_carbs'
opt_problem2 += lpSum([food_carbs[i] * amount_var[i] for i in food_names]) <= 450, 'max_carbs'
#Fiber
opt_problem2 += lpSum([food_fiber[i] * amount_var[i] for i in food_names]) >= 125, 'min_fiber'
opt_problem2 += lpSum([food_fiber[i] * amount_var[i] for i in food_names]) <= 250, 'max_fiber'
#Protein
opt_problem2 += lpSum([food_protein[i] * amount_var[i] for i in food_names]) >= 60, 'min_protein'
opt_problem2 += lpSum([food_protein[i] * amount_var[i] for i in food_names]) <= 100, 'max_protein'
#VitaminA
opt_problem2 += lpSum([food_vitA[i] * amount_var[i] for i in food_names]) >= 1000, 'min_vitA'
opt_problem2 += lpSum([food_vitA[i] * amount_var[i] for i in food_names]) <= 10000, 'max_vitA'
#VitaminC
opt_problem2 += lpSum([food_vitC[i] * amount_var[i] for i in food_names]) >= 400, 'min_vitC'
opt_problem2 += lpSum([food_vitC[i] * amount_var[i] for i in food_names]) <= 5000, 'max_vitC'
#Calcium
opt_problem2 += lpSum([food_calcium[i] * amount_var[i] for i in food_names]) >= 700, 'min_calcium'
opt_problem2 += lpSum([food_calcium[i] * amount_var[i] for i in food_names]) <= 1500, 'max_calcium'
#Iron
opt_problem2 += lpSum([food_iron[i] * amount_var[i] for i in food_names]) >= 10, 'min_iron'
opt_problem2 += lpSum([food_iron[i] * amount_var[i] for i in food_names]) <= 40, 'max_iron'


## Adding extra Constraints 
## a.	If a food is selected, then a minimum of 1/10 serving must be chosen. (Hint: now you will need two variables for each food i: whether it is chosen, and how much is part of the diet. You’ll also need to write a constraint to link them.)
for f in food_names:
    opt_problem2 += amount_var[f] <= 1000000 * binaryVars[f]
    opt_problem2 += amount_var[f] >= 0.1 * binaryVars[f]
    
## b.	Many people dislike celery and frozen broccoli. So at most one, but not both, can be selected.
opt_problem2 += binaryVars['Celery, Raw'] + binaryVars['Frozen Broccoli'] <= 1, 'At most one, but not both'

## c.	To get day-to-day variety in protein, at least 3 kinds of meat/poultry/fish/eggs must be selected. [If something is ambiguous (e.g., should bean-and-bacon soup be considered meat?), just call it whatever you think is appropriate – I want you to learn how to write this type of constraint, but I don’t really care whether we agree on how to classify foods!]
opt_problem2 += binaryVars['Roasted Chicken'] + binaryVars['Poached Eggs'] + binaryVars['Scrambled Eggs']+ \
binaryVars['Bologna,Turkey']+ binaryVars['Frankfurter, Beef'] + binaryVars['Ham,Sliced,Extralean']+ \
binaryVars['Kielbasa,Prk']+binaryVars['Pizza W/Pepperoni']+binaryVars['Pork'] + binaryVars['White Tuna in Water'] \
+binaryVars['Chicknoodl Soup'] + binaryVars['Splt Pea&Hamsoup'] + binaryVars['Vegetbeef Soup'] >= 3 , 'At least 3 proteins'

In [49]:
## Solving the optimizer
opt_problem2.solve()

1

In [50]:
#Viewing the optimal solution with this model
varsDictionary2 = {}

for v in opt_problem2.variables():
    varsDictionary2[v.name] = v.varValue
    
print(varsDictionary2)

{'Amounts_2%_Lowfat_Milk': 0.0, 'Amounts_3.3%_Fat,Whole_Milk': 0.0, 'Amounts_Apple,Raw,W_Skin': 0.0, 'Amounts_Apple_Pie': 0.0, 'Amounts_Bagels': 0.0, 'Amounts_Banana': 0.0, 'Amounts_Beanbacn_Soup,W_Watr': 0.0, 'Amounts_Bologna,Turkey': 0.0, 'Amounts_Butter,Regular': 0.0, "Amounts_Cap'N_Crunch": 0.0, 'Amounts_Carrots,Raw': 0.0, 'Amounts_Celery,_Raw': 42.399358, 'Amounts_Cheddar_Cheese': 0.0, 'Amounts_Cheerios': 0.0, 'Amounts_Chicknoodl_Soup': 0.0, 'Amounts_Chocolate_Chip_Cookies': 0.0, "Amounts_Corn_Flks,_Kellogg'S": 0.0, 'Amounts_Couscous': 0.0, 'Amounts_Crm_Mshrm_Soup,W_Mlk': 0.0, 'Amounts_Frankfurter,_Beef': 0.0, 'Amounts_Frozen_Broccoli': 0.0, 'Amounts_Frozen_Corn': 0.0, 'Amounts_Grapes': 0.0, 'Amounts_Ham,Sliced,Extralean': 0.0, 'Amounts_Hamburger_W_Toppings': 0.0, 'Amounts_Hotdog,_Plain': 0.0, 'Amounts_Kielbasa,Prk': 0.1, 'Amounts_Kiwifruit,Raw,Fresh': 0.0, 'Amounts_Lettuce,Iceberg,Raw': 82.802586, 'Amounts_Macaroni,Ckd': 0.0, 'Amounts_Malt_O_Meal,Choc': 0.0, 'Amounts_New_E_Clamch

The optimal diet after adding these constraints is below. We see that we have only Celery and Not Brocolli , which
is according to the constraints we have given. All chosen food have atleast 0.1 of the servings too.

Celery,_Raw (42.399358)
Kielbasa,Prk (0.1)
Lettuce,Iceberg,Raw (82.802586)
Oranges (3.0771841)
Peanut_Butter (1.9429716)
Poached_Eggs (0.1)
Popcorn,Air_Popped (13.223294)
Scrambled_Eggs  (0.1)